src/backend/libinput/mod.rs

```diff
@@ -193,9 +193,22 @@ impl backend::PointerMotionEvent<LibinputInputBackend> for event::pointer::Point
     fn delta_x(&self) -> f64 {
         self.dx()
     }
+
     fn delta_y(&self) -> f64 {
         self.dy()
     }
+
+    fn delta_x_unaccel(&self) -> f64 {
+        self.dx_unaccelerated()
+    }
+
+    fn delta_y_unaccel(&self) -> f64 {
+        self.dy_unaccelerated()
+    }
+
+    fn utime(&self) -> u64 {
+        event::pointer::PointerEventTrait::time_usec(self)
```

Comment:
`zwp_relative_pointer_v1::relative_motion` uses 64-bit microsecond precision timestamps while `wl_pointer::motion` uses 32-bit millisecond precision.

Looks like `input` provides `time_usec` for everything that provides `time`? So we could just expose microsecond precision timestamps and convert to millisecond. Not sure how other backends compare, though of course they can be converted even if the precision isn't there.

Reply:
Yeah that sounds sensible. So then we need to update the `time` function to return a `u64`, right?

Reply:
Right. Or some wrapper type.

Reply:
Not sure if there is much to be gained by wrapping it up tbh. A seemingly random u64 value is already unusable except for comparison.

Reply:
I don't think "a seemingly random u64 unusable except for comparison" really conforms to the protocol specification. https://wayland.app/protocols/wayland#wl_pointer:event:button says `time` is a "timestamp with millisecond granularity, with an undefined base", https://wayland.app/protocols/relative-pointer-unstable-v1#zwp_relative_pointer_v1:event:relative_motion takes "a 64 bit timestamp with microsecond granularity".

This makes it sound like a client is expected to be able to use the time information in some way other than just ordering, especially with relative pointer events (I guess something involving time between two events?). Though the protocol spec seems too vague to safely use it for... anything. It doesn't even seem to explicitly mention the time is monotonically increasing? Or if this should be the same "current time, in milliseconds, with an undefined base" used by frame callbacks?

A wrapper type could have `as_millis` and `as_micros` methods, etc. to be clearer and less error prone than `(time / 1000) as u32`, etc. Could just use `time::Duration` though that has nanosecond precision, which is unnecessary here.

Reply:
Looks like it is **not** necessarily monotonic.

> Most libinput events provide a timestamp in millisecond and/or microsecond resolution. These timestamp usually increase monotonically, but libinput does not guarantee that this always the case. In other words, it is possible to receive an event with a timestamp earlier than the previous event.

https://wayland.freedesktop.org/libinput/doc/latest/timestamps.html

Reply:
> "a 64 bit timestamp with microsecond granularity"

I would expect it to be monotonically though, even though an "unspecified base" is not mentioned explicitly. It would certainly be nice, if the protocol would explicitly mention the base no matter how it was intended. So maybe we should raise an issue on `wayland-protocols`.

Reply:
libinput's definition still sounds like it is supposed to be and a best-effort is made.

Reply:
With either `zwp_relative_pointer_v1` or with `wl_pointer`/`wl_keyboard`, the specs seem to vague to use the timestamps for much other than logging.

But I guess the way it behaves for most practical implementations is defined by what libinput does. So the same "unspecified base" should apply for `zwp_relative_pointer_v1`/`wl_pointer`/`wl_keyboard` since these would all come from the same input backend. While the "unspecified base" for frame callbacks may be different, even though the same wording is used. And the timestamps can't be guaranteed as monotonic if libinput doesn't guarantee that.

Reply:
Sounds like it would be fine to just use the same libinput value everywhere and truncate it to the required precision in that case.

Reply:
Added a commit changing `Event::time` to microseconds.